openedx_events/bridge/avro_attrs_bridge.py

```diff
@@ -1,120 +1,139 @@
+            "fields": [],
         }
 
-        record_fields = self._record_fields_for_attrs_class(attrs_cls)
-        base_schema["fields"].append(record_fields)
+        for key_name, data_type in self._signal.init_data.items():
+            base_schema["fields"].append(self._get_object_fields(key_name, data_type))
         return base_schema
 
+    def _get_object_fields(self, object_name, _object_type):
+        """
+        Create top-level field definitions for an Avro schema.
+
+        :param object_name: Field name
+        :param _object_type: (Avro) field type
+        :return:
+        """
+        inner_field = {"name": object_name}
+        if extension := self.extensions.get(_object_type, None):
+            inner_field = {
+                "name": object_name,
+                "type": extension.record_fields(),
+            }
+        # Attribute is a simple type.
+        elif _object_type in PYTHON_TYPE_TO_AVRO_MAPPING:
+            if PYTHON_TYPE_TO_AVRO_MAPPING[_object_type] in ["record", "array"]:
+                # TODO (EventBus): figure out how to handle container types. (ARCHBOM-2095)
+                # This will require better specifications for types in openedx-events repositories.
+                # This is a conversation that needs to be had with others,
+                # for now, these data types will not be supported by this bridge.
+                # TODO: this exception should be more specific.
+                raise Exception("item type for container object not specified")
+            inner_field = {
+                "name": object_name,
+                "type": PYTHON_TYPE_TO_AVRO_MAPPING[_object_type],
+            }
+
+        # _object_type is another attrs class
+        elif hasattr(_object_type, "__attrs_attrs__"):
+            # Inner Attrs Class
+
+            # fastavro does not allow you to redefine the same record type more than once,
+            # so only define an attr record once
+            if _object_type.__name__ in self.schema_record_names:
+                inner_field = {
+                    "name": object_name,
+                    "type": _object_type.__name__,
+                }
+            else:
+                self.schema_record_names.add(_object_type.__name__)
+                inner_field = self._record_fields_for_attrs_class(
+                    _object_type, object_name
+                )
+        # else _object_type is a custom type and
+        # there needs to be AvroAttrsBridgeExtension for _object_type in self.extensions
+        else:
+            raise TypeError(
+                f"Object type {_object_type} is not supported by AvroAttrsBridge. The object type needs to either"
+                " be one of the types in PYTHON_TYPE_TO_AVRO_MAPPING, an attrs decorated class, or one of the types"
+                " defined in self.extensions."
+            )
+
+        return inner_field
+
     def _record_fields_for_attrs_class(
```

Comment:
1. Given "record" can be both a noun and verb, and reads like a verb when starting a method name, can we come up with a better name?
2. The docstring mentions "Avro record" but the method mentions "record_fields". Can we make these more consistent?

Potential new names:
* `_attrs_class_to_avro_record`
* `_generate_avro_record_from_attrs_class`
* `_create_avro_record_from_attrs_class`

Note: This comment would apply to similarly named methods.